xkcd_cli/iv.py

```diff
@@ -0,0 +1,407 @@
+            cell_width, cell_height = -1, -1
+        else:
+            cell_height, cell_width = float(size_ret[1].split("=")[1]), float(
+                size_ret[2].strip("\07")
+            )
+            if len(size_ret) == 4:
+                factor = float(size_ret[3].strip("\07"))
+            else:
+                factor = 1
+            cell_height, cell_width = round(cell_height * factor), round(
+                cell_width * factor
+            )
+
+        return cell_height, cell_width
+
+    def terminal_size(self) -> Tuple[int, int]:
+        """
+        Use ANSI escape code to determine the terminal dimensions.
+        See also: https://notes.burke.libbey.me/ansi-escape-codes/
+        """
+        size_ret = self.terminal_request("\x1b[18t", "t").split(";")
+        if len(size_ret) < 3 or not size_ret[0][-1] == "8":
+            terminal_lines, terminal_cols = -1, -1
+        else:
+            terminal_lines, terminal_cols = int(size_ret[1]), int(size_ret[2][:-1])
+        return terminal_lines, terminal_cols
+
+    # Identify supported protocols
+    def have_iterm(self) -> bool:
+        """
+        Identify if the current terminal is an `iterm` terminal instance by
+        running a proprietary command that only works on those terminals.
+
+        This method sets the `iterm` variable on the current instance.
+        """
+        if self.iterm is None:
+            self.iterm = -1 not in self.iterm_cell_size()
+        return self.iterm
+
+    def have_kitty(self) -> bool:
+        """
+        Identify if the current terminal is a `kitty` terminal instance by
+        sending a sample 24-bit RGB image with dimensions 1x1 (params s, v) in
+        query action mode (a=q).
+
+        This method sets the `kitty` variable on the current instance.
+
+        See also: https://sw.kovidgoyal.net/kitty/graphics-protocol/?highlight=image#querying-support-and-available-transmission-mediums
+        """
+        if self.kitty is None:
+            self.kitty = "_G" in self.terminal_request(
+                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=24;AAAA\x1b\\", "\\"
+            )
+        return self.kitty
+
+    def have_extended_kitty(self) -> bool:
+        """
+        Determine if extended support for PNG images (f=100) in kitty is
+        available by sending a 1x1 pixel PNG image.
+
+        This method sets the `ex_kitty` variable, and the `kitty` variable if
+        unset, on the currrent instance.
+        """
+        if self.ex_kitty is None:
+            # A small jpeg from https://stackoverflow.com/questions/2253404
+            self.ex_kitty = "OK" in self.terminal_request(
+                "\x1b_Gi=31,s=1,v=1,a=q,t=d,f=100;"
+                + "/9j/2wBDAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQH/wgALCAABAAEBAREA/8QAFAABAAAAAAAAAAAAAAAAAAAAA//aAAgBAQAAAAE//9k="
+                + "\x1b\\",
+                "\\",
+            )
```

Comment:
@MatanZ For me it's still not clear how you came up with this idea. I understand that you're rendering a PNG (`f=100`) image with 1x1 dimensions `s=1,v=1` in query action mode (`a=q`). But I don't understand how you came up with this specific payload. Can you please add a comment to explain how you did this so that in case anything breaks later have a better understanding of how this works? Thanks a ton in advance.